File: src/main/java/com/portofino/realtrainmodunofficial/network/CopyRailPayload.java
package com.portofino.realtrainmodunofficial.network;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import com.portofino.realtrainmodunofficial.block.LargeRailCoreBlock;
import com.portofino.realtrainmodunofficial.blockentity.LargeRailCoreBlockEntity;
import com.portofino.realtrainmodunofficial.blockentity.RailCollisionBlockEntity;
import io.netty.buffer.ByteBuf;
import net.minecraft.core.BlockPos;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.network.handling.IPayloadContext;

/**
 * Requests a server-side clone of a placed rail so middle-click works in multiplayer too.
 */
public record CopyRailPayload(BlockPos pos) implements CustomPacketPayload {
    public static final Type<CopyRailPayload> TYPE = new Type<>(
        ResourceLocation.fromNamespaceAndPath(RealTrainModUnofficial.MODID, "copy_rail")
    );

    public static final StreamCodec<ByteBuf, CopyRailPayload> STREAM_CODEC = StreamCodec.composite(
        BlockPos.STREAM_CODEC,
        CopyRailPayload::pos,
        CopyRailPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnServer(CopyRailPayload payload, IPayloadContext context) {
        context.enqueueWork(() -> {
            Player player = context.player();
            if (player == null) {
                return;
            }

            // 中クリックは見えている collision ブロックに当たることが多いので、
            // そこから core を辿って同じ複製結果を返す。
            LargeRailCoreBlockEntity core = null;
            BlockPos corePos = payload.pos();
            if (player.level().getBlockEntity(payload.pos()) instanceof LargeRailCoreBlockEntity directCore) {
                core = directCore;
            } else if (player.level().getBlockEntity(payload.pos()) instanceof RailCollisionBlockEntity collision) {
                corePos = collision.getCorePos();
                if (player.level().getBlockEntity(corePos) instanceof LargeRailCoreBlockEntity collisionCore) {
                    core = collisionCore;
                }
            }

            if (core == null) {
                return;
            }

            ItemStack clone = LargeRailCoreBlock.createRailCloneStack(corePos, core);
            if (clone.isEmpty()) {
                return;
            }

            // サーバー側で選択中スロットを置き換えると、シングル/マルチどちらでも同じ見え方になる。
            player.getInventory().setItem(player.getInventory().selected, clone);
            player.containerMenu.broadcastChanges();
        });
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/CopyRailPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.block.LargeRailCoreBlock
import cc.mirukuneko.realtrainmodrenewed.blockentity.LargeRailCoreBlockEntity
import cc.mirukuneko.realtrainmodrenewed.blockentity.RailCollisionBlockEntity
import io.netty.buffer.ByteBuf
import net.minecraft.core.BlockPos
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.neoforged.neoforge.network.handling.IPayloadContext

data class CopyRailPayload(val pos: BlockPos) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<CopyRailPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "copy_rail")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, CopyRailPayload> = StreamCodec.composite(
            BlockPos.STREAM_CODEC,
            { payload -> payload.pos },
            ::CopyRailPayload,
        )

        @JvmStatic
        fun handleOnServer(payload: CopyRailPayload, context: IPayloadContext) {
            context.enqueueWork {
                val player = context.player()
                var corePos = payload.pos
                val direct = player.level().getBlockEntity(payload.pos)
                val core = when (direct) {
                    is LargeRailCoreBlockEntity -> direct
                    is RailCollisionBlockEntity -> {
                        corePos = direct.corePos
                        player.level().getBlockEntity(corePos) as? LargeRailCoreBlockEntity
                    }
                    else -> null
                } ?: return@enqueueWork

                val clone = LargeRailCoreBlock.createRailCloneStack(corePos, core)
                if (clone.isEmpty) {
                    return@enqueueWork
                }
                player.inventory.setItem(player.inventory.selectedSlot, clone)
                player.containerMenu.broadcastChanges()
            }
        }
    }
}
